File: java/15.3-sum.java
/*
 * @lc app=leetcode id=15 lang=java
 *
 * [15] 3Sum
 */
class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        if (nums == null || nums.length < 3) return new LinkedList<>();
        List<List<Integer>> res = new LinkedList<>();
        Arrays.sort(nums);
        for (int t = nums.length - 1; t >= 0; --t) {
            if (t < nums.length - 1 && nums[t] == nums[t+1]) continue;
            int i = 0, j = t - 1, vt = nums[t];
            while (i < j) {
                int vi = nums[i], vj = nums[j];
                int sum = vi + vj + vt;
                if (sum == 0) res.add(Arrays.asList(vi, vj, vt));
                if (sum <= 0) while (i < j && nums[++i] == vi);
                if (sum >= 0) while (i < j && nums[--j] == vj);
            }
        }
        return res;
    }
}

File: go/15.3-sum.go
package main

import "sort"

/*
 * @lc app=leetcode id=15 lang=golang
 *
 * [15] 3Sum
 */

// @lc code=start

type void struct{}

var member void

func threeSum(nums []int) [][]int {
	sort.Sort(sort.IntSlice(nums))
	res := make([][]int, 0)
	for i := 0; i < len(nums); {
		a := nums[i]
		j, k := i+1, len(nums)-1
		for j < k {
			b, c := nums[j], nums[k]
			sum := a + b + c
			if sum == 0 {
				res = append(res, []int{a, b, c})
			}
			if sum >= 0 {
				for j < k && nums[k] == c {
					k--
				}
			}
			if sum <= 0 {
				for j < k && nums[j] == b {
					j++
				}
			}
		}
		for i < len(nums) && nums[i] == a {
			i++
		}
	}
	return res
}

// @lc code=end
